Frameworks/CoreText/CTLine.mm

```diff
@@ -283,16 +283,44 @@ CGRect CTLineGetImageBounds(CTLineRef line, CGContextRef context) {
 }
 
 /**
- @Status Caveat
- @Notes Values only reflect the typographical maximums for the font - string is only evaulated
-        for width
+ @Status Interoperable
+ @Notes
 */
 double CTLineGetTypographicBounds(CTLineRef lineRef, CGFloat* ascent, CGFloat* descent, CGFloat* leading) {
-    if (!lineRef) {
+    _CTLine* line = static_cast<_CTLine*>(lineRef);
+
+    if (!line || [line->_runs count] == 0) {
         return 0;
     }
 
-    _CTLine* line = static_cast<_CTLine*>(lineRef);
+    // Created with impossible values FLT_MIN which signify they need to be populated
+    if (line->_ascent == FLT_MIN || line->_descent == FLT_MIN) {
+        for (_CTRun* run in static_cast<id<NSFastEnumeration>>(line->_runs)) {
+            DWRITE_GLYPH_METRICS glyphMetrics[run->_dwriteGlyphRun.glyphCount];
+            THROW_IF_FAILED(run->_dwriteGlyphRun.fontFace->GetDesignGlyphMetrics(run->_dwriteGlyphRun.glyphIndices,
+                                                                                 run->_dwriteGlyphRun.glyphCount,
+                                                                                 glyphMetrics,
+                                                                                 run->_dwriteGlyphRun.isSideways));
+
+            CTFontRef font = static_cast<CTFontRef>([run->_attributes objectForKey:static_cast<NSString*>(kCTFontAttributeName)]);
+            CGFloat pointSize = kCTFontSystemFontSize;
+            if (font) {
+                pointSize = CTFontGetSize(font);
+            }
+
+            DWRITE_FONT_METRICS fontMetrics;
+            run->_dwriteGlyphRun.fontFace->GetMetrics(&fontMetrics);
+            CGFloat pointsPerDesignUnit = pointSize / fontMetrics.designUnitsPerEm;
```

Comment:
I don't agree that performance in tight loop doesn't matter.  This is a case of having a helper function where one is not needed.  You need a multiplication factor, which you compute once and use everywhere.  floating point divisions are expensive and when done many times (as in this case, we would do it with every glyph) will result in "why the hell is it performing so bad".  Granted, premature optimization is bad, but this is a case where we aren't gaining anything from a helper function. 

---
In reply to: [80552492](https://github.com/Microsoft/WinObjC/pull/1027#discussion_r80552492) [](ancestors = 80552492,80534274,80529943,80528495)